src/databricks/labs/remorph/assessments/scheduler/pipeline_scheduler.py

```diff
@@ -0,0 +1,152 @@
+            should_run = True
+            logging.info(f"First time running the step: '{step.name}'.")
+        else:
+            # The Step has been run once already
+            if scheduled_frequency == ScheduledFrequency.ONCE.value:
+                logging.info(f"Step '{step.name}' has already been run once. Skipping.")
+                should_run = False
+            # Check if it's been >= 24 hours since the last run
+            elif (scheduled_frequency == ScheduledFrequency.DAILY.value and
+                  get_hours_since_last_run(last_run_time) >= 24):
+                should_run = True
+                logging.info(f"Running daily step '{step.name}' now.")
+            # Check if it's been >= 7 days since the last run
+            elif (scheduled_frequency == ScheduledFrequency.WEEKLY.value and
+                  get_days_since_last_run(last_run_time) >= 7):
+                should_run = True
+                logging.info(f"Running weekly step '{step.name}' now.")
+            else:
+                # None of the triggering frequency conditions have been met
+                should_run = False
+        return should_run
+
+    def _run_step(self, pipeline: PipelineClass, step: Step) -> str:
+        """Executes a pipeline step if it's time to run."""
+        if self._should_run(pipeline.config.name, step):
+            status = str(pipeline.execute_step(step).value)
+            self._record_run_time(pipeline.config.name, step.name, status)
+        else:
+            status = StepExecutionStatus.SKIPPED.value
+        return status
+
+    def run(self, max_num_cycles: int = None) -> list[PollingStatus]:
+        """Create an infinite loop over the pipeline steps"""
+        logging.info("PipelineScheduler has started...")
+        cycle_counter = 0
+        polling_status = []
+        while cycle_counter < max_num_cycles if max_num_cycles is not None else True:
+
+            # Loop through the list of scheduled pipelines
+            # TODO: Parallelize this in the future to be more efficient
+            for pipeline in self.pipelines:
+                pipeline_steps = pipeline.config.steps
+                for pipeline_step in pipeline_steps:
+                    # Inspect the execution frequency of the Step
+                    # Possible frequencies include: "once", "daily", "weekly" (see `ScheduledFrequency` enum)
+                    logging.info(f"Evaluating scheduling step '{pipeline_step.name}'.")
+                    step_exec_status = self._run_step(pipeline, pipeline_step)
+                    execution_status = PollingStatus(datetime.now(), pipeline.config.name,
+                                                     pipeline_step.name, step_exec_status)
+                    polling_status.append(execution_status)
+
+            # Wait a bit before polling status
+            time.sleep(self.polling_interval)
+            if max_num_cycles is not None:
+                cycle_counter += 1
```

Comment:
I would:

1. Always increment the counter, even if there's no max.
2. Include the current value of the counter in the logging.